util-secret/src/main/java/com/hpe/caf/secret/SecretUtil.java

```diff
@@ -0,0 +1,149 @@
+
+        // 1. Try environment variable
+        final String envValue = getFromEnvironment(key);
+        if (envValue != null) {
+            return envValue;
+        }
+
+        // 2. Try file reference (via environment variable)
+        final String fileValue = getFromFileViaEnvironment(key);
+        if (fileValue != null) {
+            return fileValue;
+        }
+
+        // 3. Try system property
+        final String propertyValue = getFromSystemProperty(key);
+        if (propertyValue != null) {
+            return propertyValue;
+        }
+
+        LOG.debug("No value found for key '{}' in any location", key);
+        return null;
+    }
+
+    /**
+     * Retrieves a secret value from various sources in a prescribed order of precedence:
+     * <ol>
+     *   <li>Environment variables (direct value)</li>
+     *   <li>File content (path specified by environment variable with "_FILE" suffix)</li>
+     *   <li>System properties (with "CAF." prefix)</li>
+     * </ol>
+     *
+     * For example, for a key "DATABASE_PASSWORD":
+     * <ul>
+     *   <li>First checks environment variable: DATABASE_PASSWORD</li>
+     *   <li>Then checks file path in environment variable: DATABASE_PASSWORD_FILE</li>
+     *   <li>Finally checks system property: CAF.DATABASE_PASSWORD</li>
+     * </ul>
+     *
+     * @param key The base key to look up the secret value. Must not be null.
+     * @return The secret value if found in any of the sources, or defaultValue if not found.
+     *         If found in a file, the content is trimmed of leading and trailing whitespace.
+     * @throws IOException If there is an error reading the file when using the _FILE variant
+     * @throws NullPointerException If either the key or defaultValue parameters are null
+     */
+    public static String getSecret(final String key, final String defaultValue) throws IOException
+    {
+        Objects.requireNonNull(defaultValue, "defaultValue");
+        final String value = getSecret(key);
+        if (value != null) {
+            return value;
+        } else {
+            LOG.debug("Returning default value for key '{}'", key);
+            return defaultValue;
+        }
+    }
+
+    private static String getFromEnvironment(final String key)
+    {
+        final String value = System.getenv(key);
+        if (value != null) {
+            LOG.debug("Found value for key '{}' in environment variables", key);
+            return value;
+        }
+        return null;
+    }
+
+    private static String getFromFileViaEnvironment(final String key) throws IOException
+    {
+        final String keyWithFilePostfix = key + FILE_POSTFIX;
+        final String filePath = System.getenv(keyWithFilePostfix);
+        if (filePath != null) {
+            LOG.debug("Found value for key '{}' in environment variables, reading file content...", keyWithFilePostfix);
+            final String fileContent = Files.readString(Paths.get(filePath)).trim();
```

Comment:
This will throw an exception if there is a env var ending in `_FILE`, but we are unable to read it for any reason. I wasn't sure if we wanted to be more lenient here? Perhaps logging a warning and continuing with the rest of the main logic?

Reply:
I think throwing an exception if there's an issue reading the file is fine.  But I don't understand why the `.trim()` is present.  Shouldn't a secret be allowed to end in whitespace?

Reply:
Yes thanks, `trim` removed now.